File: scripts/tf_to_pose.py
#!/usr/bin/env python
"""
Copyright 2023 Lucas Walter

Look up a transform between parent and child, then republish as a PoseStamped,
this is convenient for viewing orientation as rpy in degrees in PlotJuggler

rosrun tf_demo flyer.py
rosrun tf_demo tf_to_pose.py _parent:=map _child:=flyer

"""

import copy
from threading import Lock

import rospy
import tf2_ros
import tf2_py as tf2
from ddynamic_reconfigure_python.ddynamic_reconfigure import DDynamicReconfigure
from geometry_msgs.msg import (
    PoseStamped,
    TransformStamped,
)


class TfToPose:
    def __init__(self):
        self.lock = Lock()

        cache_time = rospy.get_param("~cache_time", 30.0)
        self.tf_buffer = tf2_ros.Buffer(cache_time=rospy.Duration(cache_time))
        self.tl = tf2_ros.TransformListener(self.tf_buffer)

        update_rate = rospy.get_param("~update_rate", 20.0)

        self.last_lookup_failed = None
        self.last_lookup_time = None

        self.config = None
        self.ddr = DDynamicReconfigure("")
        self.ddr.add_variable("lookup_time_offset", "offset the lookup time", 0.0, -10.0, 10.0)
        self.ddr.add_variable("lookup_time_offset_parent",
                              "offset the parent lookup time, need a reference frame", 0.0, -10.0, 10.0)
        self.ddr.add_variable("lookup_time_most_recent", "use the most recent tf", True)
        self.ddr.add_variable("parent", "lookup parent", "map")
        self.ddr.add_variable("child", "lookup child", "child")
        self.ddr.add_variable("reference_frame", "reference frame if lookup_time_offset_parent != 0.0", "map")

        self.ddr.start(self.config_callback)

        self.pose_pub = rospy.Publisher("tf_pose", PoseStamped, queue_size=3)

        # collect some transforms:
        try:
            rospy.sleep(1.0)
        except rospy.exceptions.ROSTimeMovedBackwardsException as ex:
            rospy.logwarn(ex)

        self.timer = rospy.Timer(rospy.Duration(1.0 / update_rate), self.update, reset=True)

    def config_callback(self, config, level):
        with self.lock:
            self.config = config
        return config

    def update(self, event):
        with self.lock:
            config = copy.deepcopy(self.config)
        if config is None:
            return

        cur_time = event.current_real
        ts = self.get_transform(config, cur_time)

        if ts is None:
            return

        if ts.header.stamp == self.last_lookup_time:
            rospy.logwarn_once(f"{ts.header.stamp.to_sec():0.3f}s {cur_time.to_sec():0.3f}s")
            return None
        self.last_lookup_time = ts.header.stamp

        rospy.logdebug_once(ts)

        ps = PoseStamped(header=ts.header)
        ps.pose.position.x = ts.transform.translation.x
        ps.pose.position.y = ts.transform.translation.y
        ps.pose.position.z = ts.transform.translation.z
        ps.pose.orientation = ts.transform.rotation

        self.pose_pub.publish(ps)

    # TODO(lucasw) put this into a library, old_tf_to_new_tf also uses it
    def get_default_transform(self) -> TransformStamped:
        ts = TransformStamped()
        ts.transform.rotation.w = 1.0
        return ts

    def get_transform(self, config, cur_time) -> TransformStamped:
        ts = self.get_default_transform()
        if config.lookup_time_most_recent:
            lookup_time = rospy.Time(0)
        else:
            lookup_time = cur_time + rospy.Duration(config.lookup_time_offset)

        if config.lookup_time_offset_parent != 0.0:
            if lookup_time == rospy.Time(0):
                rospy.logerr("need to use lookup_time_most_recent:=false")
            # TODO(lucasw) if lookup_time is rospy.Time(0) then need to get the most recent
            # available time, then add the duration
            lookup_time_parent = lookup_time + rospy.Duration(config.lookup_time_offset_parent)
        else:
            lookup_time_parent = lookup_time

        if config.parent == config.child and lookup_time_parent == lookup_time:
            rospy.loginfo_once("detected same parent and child at same stamp, returning identity")
            # TODO(lucasw) the lookup returns the same timestamp as the last lookup
            # only returns rospy.Time(0) the first pass, when doing this lookup
            # with same parent and child- so avoid that and return an identity transform
            ts = self.get_default_transform()
            ts.header.stamp = cur_time
            return ts

        try:
            text = f"'{config.parent}' {lookup_time_parent.to_sec():0.2f}s"
            text += f" to '{config.child}' {lookup_time.to_sec():0.2f}s"
            text += f", reference '{config.reference_frame}'"
            rospy.loginfo_once(text)
            if config.lookup_time_offset_parent != 0.0:
                trans = self.tf_buffer.lookup_transform_full(config.parent, lookup_time_parent,
                                                             config.child, lookup_time,
                                                             config.reference_frame,
                                                             rospy.Duration(0.15))
            else:
                trans = self.tf_buffer.lookup_transform(config.parent, config.child, lookup_time)
        except (tf2.ConnectivityException, tf2.LookupException, tf2.ExtrapolationException) as ex:
            if self.last_lookup_failed is not True:
                rospy.logwarn_throttle(2.0, f"lookup time: {lookup_time.to_sec():0.2f}")
                rospy.logwarn_throttle(2.0, ex)
                self.last_lookup_failed = True
            return None
        except rospy.exceptions.ROSTimeMovedBackwardsException as ex:
            rospy.logwarn(ex)
            return None
        if self.last_lookup_failed is not False:
            rospy.logwarn_throttle(2.0, f"now looking up {trans.header.frame_id} to {trans.child_frame_id}")
        self.last_lookup_failed = False

        ts.transform = trans.transform
        # this can't be a static transform if it is being updated all the time
        if trans.header.stamp == rospy.Time(0):
            rospy.logwarn_once("setting timestamp to current time (probably a static lookup)")
            trans.header.stamp = cur_time  # TODO(lucasw) optionally add an offset?
        ts.header = trans.header

        return ts


def main():
    rospy.init_node("tf_to_pose")
    _ = TfToPose()
    rospy.spin()


if __name__ == "__main__":
    main()

File: src/tf_to_pose.cpp
/*
Copyright 2022 Lucas Walter

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

1.
Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/


#include <ddynamic_reconfigure/ddynamic_reconfigure.h>
#include <geometry_msgs/TransformStamped.h>
#include <memory>
#include <tf_demo/copy_transform.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
#include <tf2_ros/transform_broadcaster.h>
#include <tf2_ros/transform_listener.h>
#include <string>

namespace tf_demo
{

/* Look up a transform then create a new tf child that has the same parent-child relationship
*/
class TfToPose
{
  ros::NodeHandle nh_;
  tf2_ros::Buffer tf_buffer_;
  tf2_ros::TransformListener tf_listener_;
  ros::Publisher pose_pub_;

  // double publish_rate_ = 20.0;
  // ros::Timer timer_;

  std::unique_ptr<ddynamic_reconfigure::DDynamicReconfigure> ddr_;

  CopyTransform copy_transform_;

  bool lookup_time_most_recent_ = true;
  double lookup_time_offset_ = 0.0;

  bool update()  // const ros::TimerEvent& te)
  {
    // auto cur_time = te.current_real;
    const auto cur_time = ros::Time::now();
    auto lookup_time = cur_time + ros::Duration(lookup_time_offset_);
    if (lookup_time_most_recent_) {
      lookup_time = ros::Time(0);
    }

    copy_transform_.broadcast_parent_ = copy_transform_.lookup_parent_;
    copy_transform_.broadcast_child_ = copy_transform_.lookup_child_;

    geometry_msgs::TransformStamped ts_out;
    const bool verbose = false;
    if (!copy_transform_.copyTransform(tf_buffer_, cur_time, lookup_time, ts_out, verbose)) {
      if (verbose) {
        ROS_WARN_STREAM_THROTTLE(8.0, "couldn't copy transform");
      }
      return false;
    }

    geometry_msgs::PoseStamped ps;
    ps.header = ts_out.header;
    ps.pose.position.x = ts_out.transform.translation.x;
    ps.pose.position.y = ts_out.transform.translation.y;
    ps.pose.position.z = ts_out.transform.translation.z;
    ps.pose.orientation = ts_out.transform.rotation;

    pose_pub_.publish(ps);

    return true;
  }

public:
  explicit TfToPose(const double cache_time) :
    tf_buffer_(ros::Duration(cache_time)),
    tf_listener_(tf_buffer_)
  {
    // TODO(lucasw) move into CopyTransform, make optional
    ddr_ = std::make_unique<ddynamic_reconfigure::DDynamicReconfigure>();
    ddr_->registerVariable<double>("lookup_time_offset", &lookup_time_offset_,
                                   "offset the lookup time", -10.0, 10.0);
    ddr_->registerVariable<bool>("lookup_time_most_recent", &lookup_time_most_recent_,
                                 "use the most recent tf", false, true);
    ddr_->registerVariable<std::string>("parent", &copy_transform_.lookup_parent_,
                                        std::string("lookup parent"));
    ddr_->registerVariable<std::string>("child", &copy_transform_.lookup_child_,
                                        std::string("lookup child"));
    // TODO(lucasw) support reference frame
    // ddr_->registerVariable<std::string>("reference_frame", &reference_frame_,
    //                                     std::string("reference frame"));
    ddr_->publishServicesTopics();

    pose_pub_ = nh_.advertise<geometry_msgs::PoseStamped>("tf_pose", 4);

    // collect some transforms
    ros::Duration(1.0).sleep();

    while (ros::ok()) {
      // This polls for a newer transform than the last update
      if (update()) {
        // having just updated don't need to update again immediately
        ros::Duration(0.02).sleep();
      } else {
        // sleep but for a smaller duration and maybe new transforms will arrive
        ros::Duration(0.002).sleep();
      }
    }
  }
};
}  // namespace tf_demo

int main(int argc, char** argv)
{
  ros::init(argc, argv, "tf_to_pose");
  float cache_time = 30.0;
  ros::param::get("~cache_time", cache_time);
  tf_demo::TfToPose tf_to_posef(cache_time);
  ros::spin();
}
